### PR TITLE
Port LeakSanitizer to emscripten

### PR DESCRIPTION
**Pull request by quantum5**
This seems sufficiently stable for an initial version.

The bullet, poppler, sqlite3 tests all run fine under LSan, without false positives. Leaks were introduced into sqlite3 and it was caught by LSan.

Limitations:

* Reduced execution speed, but it's much better after #8778 
* No pthread support for now

### CODE REVIEW

File: system/lib/compiler-rt/lib/lsan/lsan_emscripten.cc [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  ScanGlobalRange(0, (uptr) &__data_end, frontier);
+}
+
+LoadedModule *GetLinker() { return linker; }
+
+void ProcessPlatformSpecificAllocations(Frontier *frontier) {}
+
+// While calling Die() here is undefined behavior and can potentially
+// cause race conditions, it isn't possible to intercept exit on linux,
+// so we have no choice but to call Die() from the atexit handler.
```
Comment (tlively) [inline system/lib/compiler-rt/lib/lsan/lsan_emscripten.cc]: I assume this comment is still valid for Emscripten. s/linux/Emscripten

File: system/lib/dlmalloc.c [hunk 2]
```diff
@@ -850,6 +850,37 @@ extern "C" {
     /* ------------------- Declarations of public routines ------------------- */
     
 #ifndef USE_DL_PREFIX
+// XXX Emscripten XXX
+#if defined(__EMSCRIPTEN__) && !defined(__asmjs__)
+// XXX fastcomp HACK: in fastcomp, weak aliases loses to JavaScript versions of the function.
+// Therefore, we must define real functions. Remove this hack when fastcomp is removed.
```
Comment (tlively) [inline system/lib/dlmalloc.c]: It's not super clear to me which code this comment says is a "hack". I think it's actually the old code down below in the `#else`? It would be good to clarify that.

File: system/lib/dlmalloc.c [hunk 3]
```diff
@@ -6032,8 +6064,15 @@ int mspace_mallopt(int param_number, int value) {
...
-extern __typeof(free) emscripten_builtin_free __attribute__((weak, alias("free")));
+#ifdef __asmjs__
+extern __typeof(malloc) emscripten_builtin_malloc __attribute__((alias("malloc")));
+extern __typeof(free) emscripten_builtin_free __attribute__((alias("free")));
+extern __typeof(memalign) emscripten_builtin_memalign __attribute__((alias("memalign")));
+#else
+extern __typeof(malloc) emscripten_builtin_malloc __attribute__((alias("dlmalloc")));
+extern __typeof(free) emscripten_builtin_free __attribute__((alias("dlfree")));
+extern __typeof(memalign) emscripten_builtin_memalign __attribute__((alias("dlmemalign")));
+#endif /* __asmjs__ */
```
Comment (tlively) [inline system/lib/dlmalloc.c]: It would be good to mark these blocks with a TODO to remove with asmjs as well.

File: emcc.py [hunk 4]
```diff
@@ -1203,7 +1203,7 @@ def check(input_file):
       if shared.Settings.WASM_BACKEND:
         # setjmp/longjmp and exception handling JS code depends on this so we
         # include it by default.  Should be eliminated by meta-DCE if unused.
-        shared.Settings.EXPORTED_FUNCTIONS += ['_setThrew']
+        shared.Settings.EXPORTED_FUNCTIONS += ['_setThrew', '_emscripten_builtin_malloc', '_emscripten_builtin_free']
```
Comment (kripken) [inline emcc.py]: This doesn't look good - we shouldn't export functions without a reason. An exported function increases code size. Why is this here?

File: src/library.js [hunk 5]
```diff
@@ -931,8 +933,13 @@ LibraryManager.library = {
     name = UTF8ToString(name);
     if (!ENV.hasOwnProperty(name)) return 0;
 
+#if WASM_BACKEND
+    if (_getenv.ret) _emscripten_builtin_free(_getenv.ret);
+    _getenv.ret = allocateUTF8(ENV[name], _emscripten_builtin_malloc);
```
Comment (kripken) [inline src/library.js]: why does getenv use this instead of `_malloc`?

File: src/library_syscall.js [hunk 6]
```diff
@@ -455,7 +455,11 @@ var SyscallsLibrary = {
       FS.munmap(stream);
       SYSCALLS.mappings[addr] = null;
       if (info.allocated) {
+#if WASM_BACKEND
+        _emscripten_builtin_free(info.malloc);
```
Comment (kripken) [inline src/library_syscall.js]: why?

**Comment by kripken** [comment]
The failures on `other` and `wasm2` look real here.

**Comment by kripken** [comment]
Are the latest changes pushed to here? Aside from the failures I see `_builtin_` stuff that I thought was being removed?

**Comment by quantum5** [comment]
The `emscripten_builtin_*` functions were removed from getenv. But we still need them for `mmap` and `munmap`. This is because the memory allocator in sanitizers depends on `mmap` to hand out memory, and our `mmap` depends on `memalign`, which is overridden by the sanitizers. To avoid a cycle, we must call directly into our builtin memory allocator.

In general, any `mmap`-based memory allocator would need this change, since `mmap` is not supposed to call into user space.

I shall fix the test failures.

File: tests/test_other.py [hunk 7]
```diff
@@ -136,6 +138,20 @@ def do_other_test(self, dirname, emcc_args=[], run_args=[]):
     seen = run_js('a.out.js', args=run_args, stderr=PIPE, full_output=True) + '\n'
     self.assertContained(expected, seen)
 
+  def do_smart_test(self, source, name='test.cpp', literals=[], regexes=[],
+                    emcc_args=[], run_args=[], assert_returncode=0):
+    shutil.copyfile(source, name)
```
Comment (sbc100) [inline tests/test_other.py]: Do you need to copy here?  Can you just compile `source` directly?

File: tests/test_other.py [hunk 8]
```diff
@@ -136,6 +138,20 @@ def do_other_test(self, dirname, emcc_args=[], run_args=[]):
...
+    run_process([PYTHON, EMCC, name] + emcc_args)
+    seen = run_js('a.out.js', args=run_args, stderr=PIPE, full_output=True,
+                  assert_returncode=assert_returncode) + '\n'
+
+    for literal in literals:
+      self.assertContained([literal], seen)
+
+    for regex in regexes:
+      if not re.search(regex, seen):
+        self.fail('Expected regex "%s" to match on:\n%s' % (regex, seen))
```
Comment (sbc100) [inline tests/test_other.py]: How about `self.assertTrue(re.search(regex, seen), ..msg..)`?

File: tests/test_other.py [hunk 9]
```diff
@@ -136,6 +138,20 @@ def do_other_test(self, dirname, emcc_args=[], run_args=[]):
     seen = run_js('a.out.js', args=run_args, stderr=PIPE, full_output=True) + '\n'
     self.assertContained(expected, seen)
 
+  def do_smart_test(self, source, name='test.cpp', literals=[], regexes=[],
+                    emcc_args=[], run_args=[], assert_returncode=0):
+    shutil.copyfile(source, name)
+    run_process([PYTHON, EMCC, name] + emcc_args)
```
Comment (sbc100) [inline tests/test_other.py]: `self.get_emcc_args()` is better I think

File: tests/test_other.py [hunk 10]
```diff
@@ -136,6 +138,20 @@ def do_other_test(self, dirname, emcc_args=[], run_args=[]):
     seen = run_js('a.out.js', args=run_args, stderr=PIPE, full_output=True) + '\n'
     self.assertContained(expected, seen)
 
+  def do_smart_test(self, source, name='test.cpp', literals=[], regexes=[],
```
Comment (sbc100) [inline tests/test_other.py]: I worry a little about adding even more ways to compile and run test code.  I can't think of a suitable replacement right now.

Can you at least move it up next to `do_other_test` and document it?

File: tools/system_libs.py [hunk 11]
```diff
@@ -982,15 +982,35 @@ class libsanitizer_common_rt_wasm(CompilerRTWasmLibrary, MTLibrary):
...
   never_force = True
 
   includes = [['system', 'lib', 'compiler-rt', 'lib']]
-  cflags = ['-std=c++11', '-DUBSAN_CAN_USE_CXXABI']
+  cflags = ['-std=c++11']
+  src_glob = '*.cc'
+
+
+class libubsan_rt_wasm(SanitizerLibrary):
+  name = 'libubsan_rt_wasm'
```
Comment (sbc100) [inline tools/system_libs.py]: Do these libraries need to end in `_wasm`?   I think think that was only need when we have certain parts a library that were wasm only.  In these cases I don't think its needed.

Comment (quantum5) [reply tools/system_libs.py]: In this case, the entire library is wasm only. I think we can wait until we get rid of fastcomp to get rid of the `_wasm` suffix.

File: tests/test_other.py (hunk 7 above)
Comment (quantum5) [reply tests/test_other.py]: The copy is needed in some cases where the file name is part of the test, like in stack traces. I think it adds complexity to copy conditionally.

File: tests/test_other.py (hunk 9 above)
Comment (quantum5) [reply tests/test_other.py]: I don't think anything in `test_other.py` uses `self.emcc_args`.

File: tests/test_other.py (hunk 10 above)
Comment (quantum5) [reply tests/test_other.py]: It is right next to `do_other_test`?

File: src/library_syscall.js [hunk 12]
```diff
@@ -458,7 +458,15 @@ var SyscallsLibrary = {
       FS.munmap(stream);
       SYSCALLS.mappings[addr] = null;
       if (info.allocated) {
+#if USE_LSAN
+        if (info.anonymous) {
```
Comment (kripken) [inline src/library_syscall.js]: this is not very obvious. how about a comment explaining why we need to do this?

File: src/library_syscall.js [hunk 13]
```diff
@@ -1012,18 +1021,23 @@ var SyscallsLibrary = {
     // but it is widely used way to allocate memory pages on Linux, BSD and Mac.
     // In this case fd argument is ignored.
     if ((flags & {{{ cDefine('MAP_ANONYMOUS') }}}) !== 0) {
+#if USE_LSAN
```
Comment (kripken) [inline src/library_syscall.js]: same here, comment would help

File: src/runtime_strings.js [hunk 14]
```diff
@@ -376,9 +376,9 @@ function lengthBytesUTF32(str) {
 
 // Allocate heap space for a JS string, and write it there.
 // It is the responsibility of the caller to free() that memory.
-function allocateUTF8(str) {
+function allocateUTF8(str, allocator) {
```
Comment (kripken) [inline src/runtime_strings.js]: this increases code size, and makes me think - how about an alternative approach, where you do things like this?

```
function runWithBuiltinMalloc(func) {
  var prev = _malloc;
  _malloc = _builtin_malloc;
  try {
    func();
  } finally {
    _malloc = prev;
  }
}
```

This will put the special code in the more obvious place, perhaps, and avoid touching irrelevant code like `allocateUTF8`?

File: system/lib/compiler-rt/lib/lsan/lsan.cc [hunk 15]
```diff
@@ -59,7 +68,16 @@ static void InitializeFlags() {
   // Override from user-specified string.
   const char *lsan_default_options = MaybeCallLsanDefaultOptions();
   parser.ParseString(lsan_default_options);
+#if SANITIZER_EMSCRIPTEN
+  char *options = (char*) EM_ASM_INT({
+    return allocateUTF8(Module['LSAN_OPTIONS'] || 0,
```
Comment (kripken) [inline system/lib/compiler-rt/lib/lsan/lsan.cc]: does speed matter here? EM_JS is faster.

File: system/lib/dlmalloc.c [hunk 16]
```diff
@@ -850,6 +850,39 @@ extern "C" {
     /* ------------------- Declarations of public routines ------------------- */
     
 #ifndef USE_DL_PREFIX
+// XXX Emscripten XXX
```
Comment (kripken) [inline system/lib/dlmalloc.c]: making all these symbols weak may have side effects, are we not worried about that?

File: system/lib/emmalloc.cpp [hunk 17]
```diff
@@ -1193,9 +1193,10 @@ struct mallinfo mallinfo() {
 // in their code, and make those replacements refer to the original malloc
 // and free from this file.
 // This allows an easy mechanism for hooking into memory allocation.
-#if defined(__EMSCRIPTEN__)
-extern __typeof(malloc) emscripten_builtin_malloc __attribute__((weak, alias("malloc")));
-extern __typeof(free) emscripten_builtin_free __attribute__((weak, alias("free")));
+#if defined(__EMSCRIPTEN__) && !defined(TESTING_EMMALLOC)
```
Comment (kripken) [inline system/lib/emmalloc.cpp]: why not during testing? (comment please!)

File: system/lib/compiler-rt/lib/lsan/lsan.cc (hunk 15 above)
Comment (quantum5) [reply system/lib/compiler-rt/lib/lsan/lsan.cc]: I think this is fine. This is run exactly once on startup, so speed doesn't really matter.

File: system/lib/dlmalloc.c (hunk 16 above)
Comment (quantum5) [reply system/lib/dlmalloc.c]: They were already weak to begin with. The difference is that now, on wasm backend, we use aliases instead of renaming all the functions.

File: src/runtime_strings.js (hunk 14 above)
Comment (quantum5) [reply src/runtime_strings.js]: Done.

File: tests/test_other.py (hunk 9 above)
Comment (quantum5) [reply tests/test_other.py]: I recall that @kripken wants no tests in `test_other` to use `self.emcc_args` and the mechanism behind that.

File: tools/system_libs.py (hunk 11 above)
Comment (quantum5) [reply tools/system_libs.py]: Also note that this is already in the code. I don't think we should change the name of all sanitizer libraries in this change.

File: tests/test_other.py (hunk 7 above)
Comment (quantum5) [reply tests/test_other.py]: Actually, I think I can remove this.

File: tools/system_libs.py (hunk 11 above)
Comment (sbc100) [reply tools/system_libs.py]: OK, lets do that as a followup perhaps.

File: system/lib/compiler-rt/lib/lsan/lsan_emscripten.cc [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+// While calling Die() here is undefined behavior and can potentially
+// cause race conditions, it isn't possible to intercept exit on Emscripten,
+// so we have no choice but to call Die() from the atexit handler.
+void HandleLeaks() {
+  if (common_flags()->exitcode) Die();
+}
+
+void DoStopTheWorld(StopTheWorldCallback callback, void *argument) {
+  StopTheWorld(callback, argument);
```
Comment (tlively) [inline system/lib/compiler-rt/lib/lsan/lsan_emscripten.cc]: Comment that this is a nop please!

File: tests/test_other.py (hunk 9 above)
Comment (sbc100) [reply tests/test_other.py]: I believe that is incorrect.  For `self.set_setting()` and `self.emcc_args` shouldn't work for all test subclasses.   Thats the point of the common code in runner.py.  @kripken can you confirm?

The `set_setting`/`get_setting`/`emcc_args`/`build` stuff in RunnerCore is all generic right?
